src/Qir/Runtime/stdlib/src/arrays.rs

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::{__quantum__rt__fail, strings::convert, update_counts};
+use std::{rc::Rc, usize};
+
+#[derive(Debug, Clone)]
+pub struct QirArray {
+    pub(crate) elem_size: usize,
+    pub(crate) data: Vec<u8>,
+}
+
+#[no_mangle]
+pub extern "C" fn __quantum__rt__array_create_1d(elem_size: u32, count: u64) -> *const QirArray {
+    let elem_size = elem_size
+        .try_into()
+        .expect("The `elem_size` argument should fit in the `usize` type for this platform.");
+    let count: usize = count
+        .try_into()
+        .expect("The `count` argument should fit in the `usize` type for this platform.");
+    let data = vec![0_u8; elem_size * count];
+    Rc::into_raw(Rc::new(QirArray { elem_size, data }))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __quantum__rt__array_copy(
+    arr: *const QirArray,
+    force: bool,
+) -> *const QirArray {
+    let rc = Rc::from_raw(arr);
```

Comment:
Why not:

```suggestion
    let rc = ManuallyDrop::new(Rc::from_raw(arr));
```

Then there's no chance of forgetting `Rc::into_raw(rc)` at the end of the scope, which could accidentally free memory. (This applies to many places in the code.)

Reply:
That's a great point, and in fact makes me realize that most of the `Rc::from_raw` usage in this file (and other) places is likely not necesssary. Functions like this copy one that need to interact with the reference counts need to see it as and `Rc` can follow the pattern you suggest, and the rest can use the pointer directly without taking ownership as an `Rc`. Thanks!